File: problems/2799-count-complete-subarrays-in-an-array/solution.py
# Link: https://leetcode.cn/problems/count-complete-subarrays-in-an-array/
# Level: Medium 
# Date: 2025-04-24

class Solution:
    def countCompleteSubarrays(self, nums: List[int]) -> int:
        res = 0
        n = len(nums)
        distinct = len(set(nums))
        cnt = {}
        right = 0
        for left in range(n):
            if left > 0:
                remove = nums[left - 1]
                cnt[remove] -= 1
                if cnt[remove] == 0:
                    cnt.pop(remove)
            while right < n and len(cnt) < distinct:
                add = nums[right]
                cnt[add] = cnt.get(add, 0) + 1
                right += 1
            if len(cnt) == distinct:
                res += n - right + 1
        return res

File: problems/2799-count-complete-subarrays-in-an-array/solution.go
// Link: https://leetcode.cn/problems/count-complete-subarrays-in-an-array/
// Level: Medium
// Date: 2025-04-24

func countCompleteSubarrays(nums []int) int {
	res := 0
	cnt := make(map[int]int)
	n := len(nums)
	distinct := make(map[int]bool)
	for _, num := range nums {
		distinct[num] = true
	}
	distinctCount := len(distinct)
	right := 0
	for left := 0; left < n; left++ {
		if left > 0 {
			remove := nums[left-1]
			cnt[remove]--
			if cnt[remove] == 0 {
				delete(cnt, remove)
			}
		}
		for right < n && len(cnt) < distinctCount {
			add := nums[right]
			cnt[add]++
			right++
		}
		if len(cnt) == distinctCount {
			res += (n - right + 1)
		}
	}

	return res
}